README.md
clip-path-polygon [![Build Status](https://travis-ci.org/andrusieczko/clip-path-polygon.png?branch=master)](https://travis-ci.org/andrusieczko/clip-path-polygon)
===============

This is a jQuery plugin that makes using clip-path property easy on whatever tag under different browsers.

Tested on latest Chrome, Safari, Firefox and Opera.

It should work on Chrome ≥31, Firefox ≥31, Safari ≥7 and Opera ≥30.

What it does?

Let's say that you want to achieve something like this:

<img src="https://raw.githubusercontent.com/andrusieczko/clip-path-polygon/master/resources/example.png" width="300">

**Take a look at the [example html file](https://raw.githubusercontent.com/andrusieczko/clip-path-polygon/master/resources/example.html)!**

So you need to have a rectangle (e.g. 300x200) and you have to **crop** this image with such coordinates:
`(0; 0), (145; 0), (150; 20), (155; 0), (300; 0), (300; 200), (0; 200), (0; 0)`

### Absolute values

In *Webkit* all you have to do is to write a css style:
`-webkit-clip-path: polygon(0 0, 145px 0, 150px 20px, 155px 0, 300px 0, 300px 200px, 0 200px, 0 0)`

In *Firefox* and in W3C standard what you should do is:
`clip-path: url(#my-definition)` and somewhere in the file:

```html
<svg>
    <defs>
        <clippath id="my-definition">
            <polygon points="0 0, 145 0, 150 20, 155 0, 300 0, 300 200, 0 200, 0 0"></polygon>
        </clippath>
    </defs>
</svg>
```

### Relative values

If your design is resposive, you might want to use relative unit - percents. Then, the expected values would be:

- *Webkit*: `polygon(0 0, 49% 0, 50% 10%, 51% 0, 100% 0, 100% 100%, 0 100%, 0 0)`
- *Firefox*: `<polygon points="0 0, 0.49 0, 0.5 0.1, 0.51 0, 1 0, 1 1, 0 1, 0 0"></polygon>`

*clip-path-polygon* does this job for you!

## Installation

### Npm

Install with [https://npmjs.org](NPM) or add it to your `package.json`:

`$ npm install clip-path-polygon`

Then require it:

```javascript
require('clip-path-polygon');
```

and use:
```javascript
$myElement.clipPath();
```

### Bower

`$ bower install clip-path-polygon --save`

and then add `bower_components/clip-path-polygon/build/clip-path-polygon.min.js` to your scripts.

```html
<script src="bower_components/clip-path-polygon/build/clip-path-polygon.min.js"></script>
```

I'm assuming here that your `bower` installation folder is called `bower_components`.

### Browser

Download [clip-path-polygon.min.js (minified)](https://raw.github.com/andrusieczko/clip-path-polygon/master/build/clip-path-polygon.min.js) or [clip-path-polygon.js (dev)](https://raw.github.com/andrusieczko/clip-path-polygon/master/js/clip-path-polygon.js) and add it to you HTML file:

```html
<script src="clip-path-polygon.min.js"></script>
```

Compilation
-----------
If you want to compile the whole package with unit tests, run:
`npm install` and then `grunt` (compilation) or `grunt test` (tests).

I use [mocha](http://mochajs.org/), [sinonjs](http://sinonjs.org) and [expect.js](https://github.com/LearnBoost/expect.js) for testing.

Changelog
---------
Changlelog is available here: [CHANGELOG.md](https://github.com/andrusieczko/clip-path-polygon/blob/master/CHANGELOG.md)

API
-------

Definition:
`clipPath(points [, options])`

You can call it on a jQuery element:
```javascript
$('#my-element').clipPath(points);
```

where `points` is an array of two-elements arrays: `[[x0, y0], [x1, y1], [x2, y2]...]` crops the element to this area defined by these points.

There are some options that you can use:
<table>
  <thead>
    <th>Option</th>
    <th>Default</th>
    <th>Description</th>
  </thead>
  <tbody>
    <tr>
      <td>isPercentage</td>
      <td>*false*</td>
      <td>specifies whether you want to use absolute numbers (pixels) or relative unit (percents)</td>
    </tr>
    <tr>
      <td>isForWebkit</td>
      <td>*true*</td>
      <td>specifies if `-webkit-clip-path` property should be added to element</td>
    </tr>
    <tr>
      <td>isForSvg</td>
      <td>*true*</td>
      <td>specifies if `-clip-path` property and `<svg>` element should be added</td>
    </tr>
    <tr>
      <td>svgDefId</td>
      <td>*clipPathPolygonGenId<unique_id>*</td>
      <td>specifies *id* of SVG clippath definition</td>
    </tr>
  </tbody>
</table>

Examples
-------

### Basic

```html
<html>
  <head>
    <script src="jquery.min.js"></script>
    <script src="clip-path-polygon.min.js"></script>
    <script>
      $(function() {
        var points = [[0, 0], [145, 0], [150, 20], [155, 0], [300, 0], [300, 200], [0, 200], [0, 0]];
        $('#test').clipPath(points);
      });
    </script>
  </head>
  <body>
    <div style="width: 300px; height: 200px">
      <div id="test" style="width: 100%; height: 100%; background-color: green"></div>
    </div>
  </body>
</html>
```

which gives you such an html code (remember that `svg` element has to have *http://www.w3.org/2000/svg* namespace!):

```html
<div style="width: 300px; height: 200px">
  <div id="test" style="width: 100%; height: 100%; background-color: green;
    -webkit-clip-path: polygon(0 0, 145px 0, 150px 20px, 155px 0, 300px 0, 300px 200px, 0 200px, 0 0);
    clip-path: url(#clipPathPolygonGenId)"></div>
</div>

<svg>
  <defs>
    <clippath id="clipPathPolygonGenId">
      <polygon points="0 0, 145 0, 150 20, 155 0, 300 0, 300 200, 0 200, 0 0"></polygon>
    </clippath>
  </defs>
</svg>
```

### Relative values

To use relative values, you have to pass `isPercentage` option. You might want to name the SVG `clippath` definition differently too:

```html
<html>
  <head>
    <script src="jquery.min.js"></script>
    <script src="clip-path-polygon.min.js"></script>
    <script>
      $(function() {
        var points = [[0, 0], [49, 0], [50, 10], [51, 0], [100, 0], [100, 100], [0, 100], [0, 0]];
        $('#test').clipPath(points, {
          isPercentage: true,
          svgDefId: 'mySvg'
        });
      });
    </script>
  </head>
  <body>
    <div style="width: 300px; height: 200px">
      <div id="test" style="width: 100%; height: 100%; background-color: green"></div>
    </div>
  </body>
</html>
```

which gives you such an html code (remember that `svg` element has to have *http://www.w3.org/2000/svg* namespace!):

```html
<div style="width: 300px; height: 200px">
  <div id="test" style="width: 100%; height: 100%; background-color: green;
    -webkit-clip-path: polygon(0 0, 49% 0, 50% 10%, 51% 0, 100% 0, 100% 100%, 0 100%, 0 0);
    clip-path: url(#mySvg)"></div>
</div>

<svg>
  <defs>
    <clippath id="mySvg">
      <polygon points="0 0, 0.49 0, 0.5 0.1, 51% 0, 1 0, 1 1, 0 1, 0 0"></polygon>
    </clippath>
  </defs>
</svg>
``` 


CHANGELOG.md
# clip-path-polygon changelog

### 0.1.15 (2020-06-24)
* [UPDATE] Bump jQuery version to 3.5.0 (dependabot)

### 0.1.14 (2019-11-16)
* [UPDATE] Bump jQuery version to 3.4.1 (@elenatorro)

### 0.1.13 (2018-05-12)
* [FIX] Fix jQuery vulnerability issue by bumping the version to 3.x; update a few other dev deps

### 0.1.12 (2017-03-23)
* [BUGFIX] Webpack issue "Critical dependency: require function is used in a way in which dependencies cannot be statically extracted" fixed by @manniL

### 0.1.11 (2016-12-20)
* [BUGFIX] Chrome 55+ hot fix: render SVG first, then add clip-path property

### 0.1.10 (2016-05-07)
* [BUGFIX] change `display: none` on the created `svg` elements to `position: absolute; visibility: hidden` (sometimes `display: none` didn't work in Firefox)

### 0.1.9 (2016-05-05)
* [BUGFIX] add `window.jQuery` as an option of loading jQuery
* [BUGFIX] hide created `svg` elements so that they do not appear on the bottom of the page

### 0.1.8 (2016-02-26)
* [BUGFIX] parameters `isForWebkit`, `isForSvg` were set to false when empty map was passed options

### 0.1.7 (2016-01-01)
* [BUGFIX] parameters `isForWebkit`, `isForSvg` were ignored when `false` was passed as value
* jsdom 3.1.2 -> 7.2.2

### 0.1.6 (2015-07-21)
* [IMPROVEMENT] `svgDefId` not mandatory anymore when using more than one `.clipPath` in Firefox
* tests updated with `jsdom`

### 0.1.5 (2014-04-01)
* [FEATURE] added a percentage option (fixes [#2](https://github.com/andrusieczko/clip-path-polygon/issues/2))
* [DOCS] documentation enhaced
* [FIX] `grunt test` uses local copy of *mocha* 
* *.jshintrc* files added
* example html file provided

### 0.1.4 (2013-08-31)
* stable version working with Chrome & Firefox


test/testClipPathPolygon.js
// TODO
global.jQuery = {
  fn: function() {
  }
};

var ClipPath = require("../js/clipPathPolygon");
var sinon = require('sinon');

// TODO: packages.json
var expect = require('expect.js');

describe('ClipPath', function() {

  var cut = null;
  var jQueryMock = null;
  var $el = null;

  beforeEach(function() {
    jQueryMock = {
    find: function(selector) {
      this.selector = this.selector + " " + selector;
      return jQueryMock;
    },
    attr: function() {
      return jQueryMock;
    }
    };
    var jQuery = function(selector) {
      jQueryMock.selector = selector;
      return jQueryMock;
    };
    $el = {
      css: sinon.spy()
    };

    cut = new ClipPath(jQuery, $el, [[3, 2], [6, 7]]);
  });

  describe('_handlePxs', function() {
    
    it('should return number when withPx is false', function() {
      // given
      var expected = 20;

      // when
      var result = cut._handlePxs(20, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return number when 0 and withPx = true', function() {
      // given
      var expected = 0;

      // when
      var result = cut._handlePxs(0, true);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return number when 0 and withPx = false', function() {
      // given
      var expected = 0;

      // when
      var result = cut._handlePxs(0, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return number with pxs when number is > 0 and withPx = true', function() {
      // given
      var expected = "20px";

      // when
      var result = cut._handlePxs(20, true);

      // then
      expect(expected).to.be.equal(result);
    });
  
  });

  describe('_translatePoints', function() {

    it('should translate points with pxs', function() {
      // given
      var points = [[0, 20], [9, 2], [4, 4], [2, 0], [0, 0]];
      var expected = "0 20px, 9px 2px, 4px 4px, 2px 0, 0 0";

      // when
      var result = cut._translatePoints(points, true);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should translate points with no pxs', function() {
      // given
      var points = [[0, 20], [9, 2], [4, 4], [2, 0], [0, 0]];
      var expected = "0 20, 9 2, 4 4, 2 0, 0 0";

      // when
      var result = cut._translatePoints(points, false);

      // then
      expect(expected).to.be.equal(result);
    });

  });

  describe('create clip path definition', function() {
    it('should create clipPath using _createSvgBasedClipPath', function() {
      // given
      var id = cut.svgDefId;
      var attrSpy = sinon.spy(jQueryMock, "attr");
      var selector = "#" + id + " " + 'polygon';
      var attrArg1 = 'points';
      var attrArg2 = '2 1, 5 7, 8 8';

      // when
      cut._createSvgBasedClipPath([[2,1],[5,7],[8,8]]);

      // then
      expect(jQueryMock.selector).to.be.equal(selector);
      expect(attrSpy.withArgs(attrArg1, attrArg2).calledOnce).to.be.equal(true);
      expect($el.css.withArgs('clip-path', 'url(#' + id + ')').calledOnce).to.be.equal(true);
    });

    it('should create clipPath using _createWebkitClipPath', function() {
      // given
      var clipPathText = 'polygon(2px 1px, 0 7px, 8px 8px)';
      var attrSpy = sinon.spy(jQueryMock, "attr");

      // when
      cut._createWebkitClipPath([[2,1],[0,7],[8,8]]);

      // then
      expect(attrSpy.calledOnce).to.be.equal(false);
      expect($el.css.withArgs('-webkit-clip-path', clipPathText).calledOnce).to.be.equal(true);
    });
    
  });

});

test/test-clip-path-polygon.js
var jsdom = require('jsdom').jsdom;
document = jsdom("");
window = document.defaultView;

var ClipPath = require("../js/clip-path-polygon.js");
var sinon = require('sinon');

var expect = require('expect.js');

describe('ClipPath', function() {

  var cut = null;
  var jQueryMock = null;
  var $el = null;

  beforeEach(function() {
    jQueryMock = {
      find: function(selector) {
        this.selector = this.selector + " " + selector;
        return jQueryMock;
      },
      attr: function() {
        return jQueryMock;
      }
    };
    var jQuery = function(selector) {
      jQueryMock.selector = selector;
      return jQueryMock;
    };
    $el = {
      css: sinon.spy()
    };

    cut = new ClipPath(jQuery, $el, [
      [3, 2],
      [6, 7]
    ]);
  });

  describe('processOptions', function() {
    it('should have defaults', function() {
      // given
      expect(cut.isForWebkit).to.be.equal(true);
      expect(cut.isForSvg).to.be.equal(true);
      expect(cut.svgDefId).to.be.equal('clipPathPolygonGenId0');
      expect(cut.isPercentage).to.be.equal(false);
    });

    it ('should process passed options', function() {
      // given
      var isForWebkit = false;
      var isForSvg = false;
      var isPercentage = true;
      var svgDefId = 'definition';

      // when
      cut = new ClipPath(function() {}, $el, [], {
        isForWebkit: isForWebkit,
        isForSvg: isForSvg,
        isPercentage: isPercentage,
        svgDefId: svgDefId
      });

      // then
      expect(cut.isForWebkit).to.be.equal(isForWebkit);
      expect(cut.isForSvg).to.be.equal(isForSvg);
      expect(cut.isPercentage).to.be.equal(isPercentage);
      expect(cut.svgDefId).to.be.equal(svgDefId);
    });

    it ('should process passed options with partial options', function() {
      // given

      // when
      cut = new ClipPath(function() {}, $el, [], {});

      // then
      expect(cut.isForWebkit).to.be.equal(true);
      expect(cut.isForSvg).to.be.equal(true);
      expect(cut.isPercentage).to.be.equal(false);
      expect(cut.svgDefId).to.contain('clipPathPolygonGenId');
    });
  });

  describe('_handlePxs', function() {

    it('should return number when withUnit is false', function() {
      // given
      var expected = 20;

      // when
      var result = cut._handlePxs(20, false, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return number when 0 and withUnit = true', function() {
      // given
      var expected = 0;

      // when
      var result = cut._handlePxs(0, true, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return number when 0 and withUnit = false', function() {
      // given
      var expected = 0;

      // when
      var result = cut._handlePxs(0, false, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return number with pxs when number is > 0 and withUnit = true', function() {
      // given
      var expected = "20px";

      // when
      var result = cut._handlePxs(20, true, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return a number/100 when no unit and percentage', function() {
      // given
      var expected = 0.2;

      // when
      var result = cut._handlePxs(20, false, true);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should return a number with percent sign when  withUnit and percentage', function() {
      // given
      var expected = '20%';

      // when
      var result = cut._handlePxs(20, true, true);

      // then
      expect(expected).to.be.equal(result);
    });

  });

  describe('_translatePoints', function() {

    it('should translate points with pxs', function() {
      // given
      var points = [
        [0, 20],
        [9, 2],
        [4, 4],
        [2, 0],
        [0, 0]
      ];
      var expected = "0 20px, 9px 2px, 4px 4px, 2px 0, 0 0";

      // when
      var result = cut._translatePoints(points, true, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should translate points with no pxs', function() {
      // given
      var points = [
        [0, 20],
        [9, 2],
        [4, 4],
        [2, 0],
        [0, 0]
      ];
      var expected = "0 20, 9 2, 4 4, 2 0, 0 0";

      // when
      var result = cut._translatePoints(points, false, false);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should translate points with percents', function() {
      // given
      var points = [
        [0, 20],
        [9, 2],
        [4, 4],
        [2, 0],
        [0, 0]
      ];
      var expected = "0 20%, 9% 2%, 4% 4%, 2% 0, 0 0";

      // when
      var result = cut._translatePoints(points, true, true);

      // then
      expect(expected).to.be.equal(result);
    });

    it('should translate points with no percents and divided by 100', function() {
      // given
      var points = [
        [0, 20],
        [9, 2],
        [4, 4],
        [2, 0],
        [0, 0]
      ];
      var expected = "0 0.2, 0.09 0.02, 0.04 0.04, 0.02 0, 0 0";

      // when
      var result = cut._translatePoints(points, false, true);

      // then
      expect(expected).to.be.equal(result);
    });

  });

  describe('create clip path definition', function() {
    it('should create clipPath using _createSvgBasedClipPath', function() {
      // given
      var id = cut.svgDefId;
      var attrSpy = sinon.spy(jQueryMock, "attr");
      var selector = "#" + id + " " + 'polygon';
      var attrArg1 = 'points';
      var attrArg2 = '2 1, 5 7, 8 8';

      // when
      cut._createSvgBasedClipPath([
        [2, 1],
        [5, 7],
        [8, 8]
      ]);

      // then
      expect(jQueryMock.selector).to.be.equal(selector);
      expect(attrSpy.withArgs(attrArg1, attrArg2).calledOnce).to.be.equal(true);
      expect($el.css.withArgs('clip-path', 'url(#' + id + ')').calledOnce).to.be.equal(true);
    });

    it('should create clipPath using _createSvgBasedClipPath with percentage', function() {
      // given
      cut.isPercentage = true;
      var id = cut.svgDefId;
      var attrSpy = sinon.spy(jQueryMock, "attr");
      var selector = "#" + id + " " + 'polygon';
      var attrArg1 = 'points';
      var attrArg2 = '0.02 0.01, 0.05 0.07, 0.08 0.08';

      // when
      cut._createSvgBasedClipPath([
        [2, 1],
        [5, 7],
        [8, 8]
      ]);

      // then
      expect(jQueryMock.selector).to.be.equal(selector);
      expect(attrSpy.withArgs(attrArg1, attrArg2).calledOnce).to.be.equal(true);
      expect($el.css.withArgs('clip-path', 'url(#' + id + ')').calledOnce).to.be.equal(true);
    });

    it('should create clipPath using _createWebkitClipPath', function() {
      // given
      var clipPathText = 'polygon(2px 1px, 0 7px, 8px 8px)';
      var attrSpy = sinon.spy(jQueryMock, "attr");

      // when
      cut._createWebkitClipPath([
        [2, 1],
        [0, 7],
        [8, 8]
      ]);

      // then
      expect(attrSpy.calledOnce).to.be.equal(false);
      expect($el.css.withArgs('-webkit-clip-path', clipPathText).calledOnce).to.be.equal(true);
    });

    it('should create clipPath using _createWebkitClipPath with percentage', function() {
      // given
      cut.isPercentage = true;
      var clipPathText = 'polygon(2% 1%, 0 7%, 8% 8%)';
      var attrSpy = sinon.spy(jQueryMock, "attr");

      // when
      cut._createWebkitClipPath([
        [2, 1],
        [0, 7],
        [8, 8]
      ]);

      // then
      expect(attrSpy.calledOnce).to.be.equal(false);
      expect($el.css.withArgs('-webkit-clip-path', clipPathText).calledOnce).to.be.equal(true);
    });

  });

});

Gruntfile.js
/* global module:true */
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    uglify: {
      options: {
        banner: '/*!\n' + ' * jQuery <%= pkg.name %> Plugin v<%= pkg.version %> (<%= grunt.template.today("yyyy-mm-dd") %>)\n' + ' * <%= pkg.description %>\n' + ' * <%= pkg.url %>\n' + ' * \n' + ' * Copyright <%= grunt.template.today("yyyy") %> <%= pkg.author %>\n' + ' * Released under <%= pkg.license %> license\n' + ' */\n'
      },
      build: {
        src: '<%= pkg.main %>.js',
        dest: 'build/<%= pkg.name %>.min.js'
      }
    },
    shell: {
      jsUnitTests: {
        command: 'node node_modules/mocha/mocha.js test/**.js',
        options: {
          stdout: true,
          failOnError: true
        }
      }
    },
    mochacov: {
      coverage: {
        options: {
          coveralls: {
            serviceName: 'travis-ci'
          }
        }
      },
      test: {
        options: {
          reporter: 'spec'
        }
      },
      options: {
        files: 'test/*.js'
      }
    }
  });

  grunt.loadNpmTasks('grunt-shell');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-mocha-cov');

  grunt.registerTask('travis', ['mochacov:coverage']);

  grunt.registerTask('default', ['uglify', 'shell:jsUnitTests']);
  grunt.registerTask('test', ['shell:jsUnitTests', 'mochacov:test']);

};

js/clipPathPolygon.js
/*!
 * jQuery clipPathPolygon Plugin v0.9
 * https://github.com/andrusieczko/jquery-clippath-polygon
 * jQuery plugin that makes easy to use clip-path on whatever tag under different browsers
 *
 * Copyright 2013 Karol Andrusieczko
 * Released under the MIT license
 */
(function($) {

  var ClipPath = function(jQuery, $el, points, options) {
    this.$ = jQuery;
    this.$el = $el;
    this.points = points;

    this.processOptions(options);
  };

  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = ClipPath;
    }
    exports.ClipPath = ClipPath;
  } else {
    var globalVariable = window || root;
    globalVariable.ClipPath = ClipPath;
  }

  ClipPath.prototype = {

    $: null,
    $el: null,
    points: null,

    isForWebkit: true,
    isForSvg: true,
    svgDefId: 'clipPathPolygonGenId',

    create: function() {
      this._createClipPath(this.points);
    },

    _createClipPath: function(points) {
      this._createSvgDefs();
      if (this.isForWebkit) {
        this._createWebkitClipPath(points);
      }
      if (this.isForSvg) {
        this._createSvgBasedClipPath(points);
      }
    },

    _createWebkitClipPath: function(points) {
      var clipPath = "polygon(" + this._translatePoints(points, true) + ")";
      this.$el.css('-webkit-clip-path', clipPath);
    },

    _createSvgBasedClipPath: function(points) {
      this.$('#' + this.svgDefId + '').find('polygon').attr('points', this._translatePoints(points, false));
      this.$el.css('clip-path', 'url(#' + this.svgDefId + ')');
    },


    _translatePoints: function(points, withPxs) {
      var result = [];
      for (var i in points) {
        var x = this._handlePxs(points[i][0], withPxs);
        var y = this._handlePxs(points[i][1], withPxs);
        result.push(x + ' ' + y);
      }
      return result.join(', ');
    },

    _handlePxs: function(number, withPxs) {
      if (number === 0 || !withPxs) {
        return number;
      }
      return number + "px";
    },

    _createSvgElement: function(elementName) {
      return this.$(document.createElementNS('http://www.w3.org/2000/svg', elementName));
    },

    _createSvgDefs: function() {
      if (this.$('#' + this.svgDefId + '').size() === 0) {
        var $svg = this._createSvgElement('svg').attr('width', 0).attr('height', 0);
        var $defs = this._createSvgElement('defs');
        $svg.append($defs);
        var $clippath = this._createSvgElement('clipPath').attr('id', this.svgDefId);
        $defs.append($clippath);
        var $polygon = this._createSvgElement('polygon');
        $clippath.append($polygon);
        this.$('body').append($svg);
      }
    },

    processOptions: function(options) {
      this.isForWebkit = (options && options.isForWebkit) || this.isForWebkit;
      this.isForSvg = (options && options.isForSvg) || this.isForSvg;
      this.svgDefId = (options && options.svgDefId) || this.svgDefId;
    }
  };

  $.fn.clipPath = function(points, options) {
    return this.each(function() {
      var $el = $(this);
      var clipPath = new ClipPath($, $el, points, options);
      clipPath.create();
    });
  };

}).call(this, jQuery);

js/clip-path-polygon.js
/*!
 * jQuery clip-path-polygon Plugin v0.1.15 (2020-06-24)
 * jQuery plugin that makes easy to use clip-path on whatever tag under different browsers
 * https://github.com/andrusieczko/clip-path-polygon
 * 
 * Copyright 2020 Karol Andrusieczko
 * Released under MIT license
 */

var globalVariable = window || root;
var jQuery = jQuery || globalVariable.jQuery || require("jquery");

(function($) {
  var id = 0;

  var ClipPath = function(jQuery, $el, points, options) {
    this.$ = jQuery;
    this.$el = $el;
    this.points = points;
    this.svgDefId = 'clipPathPolygonGenId' + id++;

    this.processOptions(options);
  };

  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = ClipPath;
    }
    exports.ClipPath = ClipPath;
  } else {
    globalVariable.ClipPath = ClipPath;
  }

  ClipPath.prototype = {

    $: null,
    $el: null,
    points: null,

    isForWebkit: true,
    isForSvg: true,
    svgDefId: null,
    isPercentage: false,

    create: function() {
      this._createClipPath(this.points);
    },

    _createClipPath: function(points) {
      this._createSvgDefs();
      if (this.isForSvg) {
        this._createSvgBasedClipPath(points);
      }
      if (this.isForWebkit) {
        this._createWebkitClipPath(points);
      }
    },

    _createWebkitClipPath: function(points) {
      var clipPath = "polygon(" + this._translatePoints(points, true, this.isPercentage) + ")";
      this.$el.css('-webkit-clip-path', clipPath);
    },

    _createSvgBasedClipPath: function(points) {
      this.$('#' + this.svgDefId + '').find('polygon').attr('points', this._translatePoints(points, false, this.isPercentage));
      this.$el.css('clip-path', 'url(#' + this.svgDefId + ')');
    },


    _translatePoints: function(points, withUnit, isPercentage) {
      var result = [];
      for (var i in points) {
        var x = this._handlePxs(points[i][0], withUnit, isPercentage);
        var y = this._handlePxs(points[i][1], withUnit, isPercentage);
        result.push(x + ' ' + y);
      }
      return result.join(', ');
    },

    _handlePxs: function(number, withUnit, isPercentage) {
      if (number === 0) {
        return number;
      }
      if (!withUnit) {
        if (isPercentage) {
          return number / 100;
        }
        return number;
      }

      return number + (isPercentage ? "%" : "px");
    },

    _createSvgElement: function(elementName) {
      return this.$(document.createElementNS('http://www.w3.org/2000/svg', elementName));
    },

    _createSvgDefs: function() {
      if (this.$('#' + this.svgDefId + '').length === 0) {
        var $svg = this._createSvgElement('svg').attr('width', 0).attr('height', 0).css({
          'position': 'absolute',
          'visibility': 'hidden',
          'width': 0,
          'height': 0
        });
        var $defs = this._createSvgElement('defs');
        $svg.append($defs);
        var $clippath = this._createSvgElement('clipPath').attr('id', this.svgDefId);
        if (this.isPercentage) {
          $clippath.get(0).setAttribute('clipPathUnits', 'objectBoundingBox');
        }
        $defs.append($clippath);
        var $polygon = this._createSvgElement('polygon');
        $clippath.append($polygon);
        this.$('body').append($svg);
      }
    },

    processOptions: function(options) {
      this.isForWebkit = (options && typeof(options.isForWebkit) !== "undefined") ? options.isForWebkit : this.isForWebkit;
      this.isForSvg = (options && typeof(options.isForSvg) !== "undefined") ? options.isForSvg : this.isForSvg;
      this.isPercentage = (options && options.isPercentage || this.isPercentage);
      this.svgDefId = (options && options.svgDefId) || this.svgDefId;
    }
  };
  
  $.fn.clipPath = function(points, options) {
    return this.each(function() {
      var $el = $(this);
      var clipPath = new ClipPath($, $el, points, options);
      clipPath.create();
    });
  };

}).call(this, jQuery);
